tfhe/src/shortint/list_compression/compression.rs

```diff
@@ -0,0 +1,243 @@
+
+        assert!(
+            message_modulus.0 <= carry_modulus.0,
+            "GLWE packing is implemented with messages in carries, so carry_modulus (={}) can't be bigger than message_modulus(={})",
+            carry_modulus.0,
+            message_modulus.0 ,
+        );
+
+        let glwe_ct_list: Vec<_> = ciphertexts
+            .par_chunks(lwe_per_glwe.0)
+            .map(|ct_list| {
+                let mut list: Vec<_> = vec![];
+
+                for ct in ct_list {
+                    assert!(ct.carry_is_empty());
+
+                    assert_eq!(
+                        lwe_size,
+                        ct.ct.lwe_size(),
+                        "All ciphertexts do not have the same lwe size as the packing keyswitch key"
+                    );
+
+                    assert_eq!(
+                        message_modulus, ct.message_modulus,
+                        "All ciphertexts do not have the same message modulus"
+                    );
+                    assert_eq!(
+                        carry_modulus, ct.carry_modulus,
+                        "All ciphertexts do not have the same carry modulus"
+                    );
+                    assert_eq!(
+                        pbs_order, ct.pbs_order,
+                        "All ciphertexts do not have the same pbs order"
+                    );
+
+                    let mut ct = ct.clone();
+
+                    unchecked_scalar_mul_assign(&mut ct, message_modulus.0 as u8);
+
+                    list.extend(ct.ct.as_ref());
+                }
+
+                let list = LweCiphertextList::from_container(list, lwe_size, ciphertext_modulus);
+
+                let bodies_count = ct_list.len();
+
+                let mut out =
+                    GlweCiphertext::new(0, glwe_size, polynomial_size, ciphertext_modulus);
+
+                par_keyswitch_lwe_ciphertext_list_and_pack_in_glwe_ciphertext(
+                    lwe_pksk, &list, &mut out,
+                );
+
+                CompressedModulusSwitchedGlweCiphertext::compress(
+                    &out,
+                    self.storage_log_modulus,
+                    bodies_count,
+                )
+            })
+            .collect();
+
+        CompressedCiphertextList {
+            modulus_switched_glwe_ciphertext_list: glwe_ct_list,
+            message_modulus,
+            carry_modulus,
+            pbs_order,
+            lwe_per_glwe,
+            count,
+            ciphertext_modulus,
+        }
+    }
+}
+
+impl DecompressionKey {
+    pub fn unpack(&self, packed: &CompressedCiphertextList, index: usize) -> Option<Ciphertext> {
+        let carry_extract = generate_lookup_table(
+            self.out_glwe_size(),
+            self.out_polynomial_size(),
+            packed.ciphertext_modulus,
+            packed.message_modulus,
+            packed.carry_modulus,
+            |x| x / packed.message_modulus.0 as u64,
+        );
+        if index >= packed.count {
```

Comment:
minor thing but the if could be moved before generating the LUT

Reply:
Done